lib/ivis_opengl/gfx_api_gl.cpp

```diff
@@ -62,6 +92,38 @@ struct gl_texture : public gfx_api::texture
 
 };
 
+struct gl_buffer : public gfx_api::buffer
+{
+	gfx_api::buffer::usage usage;
+	GLuint buffer;
+
+	virtual ~gl_buffer() override
+	{
+		glDeleteBuffers(1, &buffer);
+	}
+
+	gl_buffer(const gfx_api::buffer::usage& usage, const size_t& size, const gfx_api::context::buffer_storage_hint& hint)
+		: usage(usage)
+	{
+		glGenBuffers(1, &buffer);
+		glBindBuffer(to_gl(usage), buffer);
+		glBufferData(to_gl(usage), size, nullptr, to_gl(hint));
+	}
+
+	void bind() override
+	{
+		glBindBuffer(to_gl(usage), buffer);
+	}
+
+
+	virtual void upload(const size_t & start, const size_t & size, const void * data) override
+	{
+		glBindBuffer(to_gl(usage), buffer);
+		glBufferSubData(to_gl(usage), start, size, data);
```

Comment:
`upload` accepts any size, but yet the underlying `glBufferSubData` function states:

> `glBufferSubData` and `glNamedBufferSubData` redefine some or all of the data store for the specified buffer object. Data starting at byte offset `offset` and extending for `size` bytes is copied to the data store from the memory pointed to by `data`. **`offset` and `size` must define a range lying entirely within the buffer object's data store.**

Thus passing in a size or start that are out of range is invalid - OpenGL does not automatically reallocate / extend the underlying buffer.

To make this function's implementation safe:
- `gfx_api::buffer::upload` must bounds-check `start` & `size`
    - (`glGetBufferParameteriv` with parameter `GL_BUFFER_SIZE` can be used to get the size of the current buffer, but it may be faster to store this inside the `gfx_api::buffer` struct)
    - if necessary:
        1. allocate a new buffer (`glGenBuffers`)
        2. bind the old buffer to `GL_COPY_READ_BUFFER`
        3. bind the new buffer to `GL_COPY_WRITE_BUFFER`
        4. use `glCopyBufferSubData` to copy from the old buffer to the new
        5. unbind both buffers from GL_COPY_READ and GL_COPY_WRITE (`glBindBuffer(param, 0)`)
        6. delete the old buffer (`glDeleteBuffers(1, &buffer)`)
        7. set the internal `GLuint buffer` to the new buffer
        8. *then* you can use both lines of code above to copy the original data passed into `upload` to the new buffer (which is now `buffer`)